### ISSUE TITLE
U_ARGUMENT_TYPE_MISMATCH

### ISSUE THREAD

**Issue by frent** [cakephp/cakephp#12592]
This is a (multiple allowed):

* [x] bug

* CakePHP Version: "cakephp/cakephp": "~3.4",

### What you did
Changed from PHP 7.0 to 7.1

### What happened
```
Error in Debugkit or any other php form where you add a parameter to a translation:
Normal translations without any parameter work fine.

[Aura\Intl\Exception\CannotFormat] U_ARGUMENT_TYPE_MISMATCH in /home/rlvpr/public_html/vendor/cakephp/cakephp/src/I18n/Formatter/IcuFormatter.php on line 95
Request URL: /favicon.ico
Referer URL: https://rlvpr.be/nl
Client IP: 81.82.231.253
Stack Trace:
#0 /home/rlvpr/public_html/vendor/cakephp/cakephp/src/I18n/Formatter/IcuFormatter.php(52): Cake\I18n\Formatter\IcuFormatter->_formatMessage('nl', '{0,number,#,###...', Array)
#1 /home/rlvpr/public_html/vendor/cakephp/cakephp/src/I18n/Translator.php(159): Cake\I18n\Formatter\IcuFormatter->format('nl', '{0,number,#,###...', Array)
#2 /home/rlvpr/public_html/vendor/cakephp/cakephp/src/I18n/functions.php(88): Cake\I18n\Translator->translate('{0,number,#,###...', Array)
#3 /home/rlvpr/public_html/vendor/cakephp/cakephp/src/I18n/Number.php(92): __d('cake', '{0,number,#,###...', 7.5918045043945)
#4 /home/rlvpr/public_html/vendor/cakephp/debug_kit/src/Panel/TimerPanel.php(125): Cake\I18n\Number::toReadableSize(7960584)
#5 /home/rlvpr/public_html/vendor/cakephp/debug_kit/src/ToolbarService.php(199): DebugKit\Panel\TimerPanel->summary()
#6 /home/rlvpr/public_html/vendor/cakephp/debug_kit/src/Middleware/DebugKitMiddleware.php(53): DebugKit\ToolbarService->saveData(Object(Cake\Http\ServerRequest), Object(Cake\Http\Response))
#7 /home/rlvpr/public_html/vendor/cakephp/cakephp/src/Http/Runner.php(65): DebugKit\Middleware\DebugKitMiddleware->__invoke(Object(Cake\Http\ServerRequest), Object(Cake\Http\Response), Object(Cake\Http\Runner))
#8 /home/rlvpr/public_html/vendor/cakephp/cakephp/src/Http/Runner.php(51): Cake\Http\Runner->__invoke(Object(Cake\Http\ServerRequest), Object(Cake\Http\Response))
#9 /home/rlvpr/public_html/vendor/cakephp/cakephp/src/Http/Server.php(80): Cake\Http\Runner->run(Object(Cake\Http\MiddlewareQueue), Object(Cake\Http\ServerRequest), Object(Cake\Http\Response))
#10 /home/rlvpr/public_html/webroot/index.php(38): Cake\Http\Server->run()
#11 {main}
```

### What you expected to happen
String to be populated with parameter value

**Comment by ankitpokhrel** [comment]
Suddenly getting similar error after upgrading to php 7.1

Cakephp version: 3.3.15 

```php
Fatal error: [Aura\Intl\Exception\CannotFormat] U_ARGUMENT_TYPE_MISMATCH 
#0 /home/project/public_html/vendor/cakephp/cakephp/src/I18n/Formatter/IcuFormatter.php(52): Cake\I18n\Formatter\IcuFormatter->_formatMessage('en_US', 'Render {0}', Array) 
#1 /home/project/public_html/vendor/cakephp/cakephp/src/I18n/Translator.php(150): Cake\I18n\Formatter\IcuFormatter->format('en_US', 'Render {0}', Array) 
#2 /home/project/public_html/vendor/cakephp/cakephp/src/I18n/functions.php(84): Cake\I18n\Translator->translate('Render {0}', Array) #3 /home/project/public_html/vendor/cakephp/debug_kit/src/Panel/TimerPanel.php(83): __d('debug_kit', 'Render {0}', '/home/caspaonli...') 
#4 /home/project/public_html/vendor/cakephp/cakephp/src/Event/EventManager.php(424): DebugKit\Panel\TimerPanel->DebugKit\Panel\{closure}(Object(Cake\Event\Event), '/home/project...') 
#5 /home/project/public_html/vendor/cakephp/cakephp/src/Event/EventManager.php(391): Cake\Event\EventManager->_callListener(Object(Closure), Object(Cake\Event\E in /home/project/public_html/vendor/cakephp/cakephp/src/Error/ErrorHandler.php on line 156
```

**Comment by garas** [comment]
@frent @ankitpokhrel
What is your version of ICU library? Look at `phpinfo()` at `intl` section.
What OS? What repo is PHP from (OS default, Remi, IUS, etc...)

Maybe there is some correlation

**Comment by frent** [comment]
Internationalization support | enabled
version | 1.1.0
ICU version | 4.2.1

PHP Version 7.1.22
Linux cloud.tbhost.be 2.6.32-696.18.7.el6.x86_64 #1 SMP Thu Jan 4 17:31:22 UTC 2018 x86_64

**Comment by ankitpokhrel** [comment]
I am using same ICU and PHP version as @frent

**Comment by frent** [comment]
The problem also occurs with:
PHP Version 7.2.10
Intl version | 1.1.0
ICU version | 4.2.1
ICU TZData version | 2009j
ICU Unicode version | 5.1

But not with:
PHP Version 7.0.32

Internationalization support | enabled
version | 1.1.0
ICU version | 4.2.1

So ICU version seems to be the same for all...

**Comment by ADmad** [comment]
ICU 4.2.1 is ancient. Updating your ICU libs to a more recent version should fix the problem. I don't think there's anything we can do here.

**Comment by BassemN** [comment]
I have the same issue after one of my clients upgrades to PHP 7.1.22

**Comment by BassemN** [comment]
I have the same issue after one of my clients' upgrades from PHP 7.1.20 to PHP 7.1.22 but after downgrading to PHP 5.6, the problem fixed. I think there is a compatibility issue with 7.1.22

**Comment by garas** [comment]
It is possible that PHP 7.1 became incompatible with very old ICU, even if http://php.net/manual/en/intl.requirements.php says it supports version 4.0.0 or newer.

Try to execute this script https://3v4l.org/BM5iO, and if it also fails, you can report issue to PHP bug tracker

**Comment by BassemN** [comment]
@garas 
I run the provided code and got the follwng error by the way my client ICU version is `4.2.1`
```log
[27-Sep-2018 09:33:00 UTC] PHP Fatal error:  Uncaught Exception: U_ARGUMENT_TYPE_MISMATCH in /webroot/info.php:16
Stack trace:
#0 {main}
  thrown in /webroot/info.php on line 16
```

**Comment by garas** [comment]
This means there is no fault of CakePHP.

The only way is to upgrade `ICU`

**Comment by BassemN** [comment]
The issue is that most of my clients are on shared hosting which will be difficult to force the hosting companies to upgrade the `ICU`

**Comment by garas** [comment]
Contacting hosting company is the best solution.

You can also try https://github.com/symfony/polyfill-intl-messageformatter, but because `ext-intl` is installed, you can't replace `MessageFormatter` class. 
However it can be done in conjunction with `src/I18n/Formatter/IcuFormatter.php`

**Comment by ADmad** [comment]
@BassemN What do you expect CakePHP to do if PHP 7.1 is incompatible with ICU 4.2? Take up the matter with PHP devs or find a better hosting provider.

**Comment by frent** [comment]
I do agree this is not a problem that CakePHP should fix - but users are aware now that this problem could be fixed by upgrading their ICU or change their php version.

**Comment by BassemN** [comment]
@ADmad Yes, I know this is not a CakePHP issue after applying @garas script . I just was sharing the shared hosting issue with you. I'll try to reach PHP devs.
@garas Thanks for your suggestion.

**Comment by garas** [comment]
Issue was fixed in PHP and will be in upcoming 7.1.24 and 7.2.12

- intl:
   Fixed [bug #76942](https://bugs.php.net/bug.php?id=76942) (U_ARGUMENT_TYPE_MISMATCH). (anthrax at unixuser dot org)

**Comment by BassemN** [comment]
@garas Yes, I'm the one who reported the bug to the PHP devs by using your script https://3v4l.org/BM5iO
Thanks allot @garas :heart:
